### ISSUE TITLE
Function "filter" fails when the type of the first argument is not  bool

### ISSUE THREAD

**Issue by xiaxinmeng** [exaloop/codon#241]
The function filter fails when the type of the first argument is not  bool.  

test.py
```
L = (4,9,3,14,7,10,6,1)
def is_odd(x):
    return x % 2 
newlist = list(filter(is_odd,L))
print(newlist)
```

Reproduce:
codon/codon-linux-x86_64/codon-deploy/bin/codon' run -release test.py

Crash message:
```
48793_1093447229__0.py:15:16-22: error: 'filter[is_odd[...],Generator[int],int]' does not match expected type 'filter[Callable[[int],bool],Generator[int],int]'
corrupted double-linked list
Aborted (core dumped)

```
Behavior on CPython 3.10.8: 
`[9, 3, 7, 1] `

Environment:
codon: v0.15.5 on Feb 6
Ubuntu 18.04

**Comment by Michael-F-Ellis** [comment]
Python filter functions should return a bool.  The following works as expected.
```python
L = (4,9,3,14,7,10,6,1)
def is_odd(x):
    return x % 2 > 0
newlist = list(filter(is_odd,L))
print(newlist)
```

**Comment by inumanag** [comment]
Crash is fixed in #604. Thank you @Michael-F-Ellis for the correct solution!
